soroban-env-host/src/storage.rs

```diff
@@ -193,9 +193,12 @@ impl Storage {
                 // In recording mode we treat the map as a cache
                 // that misses read-through to the underlying src.
                 if !self.map.contains_key::<LedgerKey>(key, budget)? {
+                    let val = src
+                        .get(key)
+                        .map_err(|_| HostError::from(ScHostStorageErrorCode::MissingKeyInGet))?;
```

Comment:
Losing the original error here seems not ideal. Could we take the original error, generate text for it by using it's Display impl, and add that as a DebugEvent to the returned HostError?

Reply:
Hmm I agree it would be nice to retain the original error. 
It would be difficult because the storage does not own a host and it needs it to add a DebugEvent. 
Changing the `Storage` to also return a `Box<dyn Error>` might be the more plausible way but that requires bigger change in the host. I will experiment with that.

Reply:
@leighmcculloch I have updated both `Storage` and `Footprint` to return `Box<dyn Error>` such that the upstream error is preserved. 
Also on the host side, I added the new functionality to map a `Error` into a `HostError`:  
- If the original error was a `HostError`, retrieve it. 
- Otherwise, it creates a `Bytes` object from its `fmt` string, and creates a debug event from it. (The `Bytes` representation can be improved later with better types, e.g. https://github.com/stellar/rs-soroban-env/issues/463)
Would love to have @graydon's opinion as well.